src/state.rs

```diff
@@ -365,78 +370,126 @@ impl TabExpandedString {
     }
 }
 
-/// Estimate the number of seconds per step
+/// Estimate the number of steps per second
+///
+/// Exponentially smoothed estimate with an adjustable weighting
 ///
-/// Ring buffer with constant capacity. Used by `ProgressBar`s to display `{eta}`,
-/// `{eta_precise}`, and `{*_per_sec}`.
+/// The weight used for smoothing can be changed with `set_exponent_weight_secs`. All data in the
+/// Estimator older than this value will receive 10% of the weight in the estimate. Data newer than
+/// this value will receive 90% of the weight. The default value is 15 seconds.
 pub(crate) struct Estimator {
-    steps: [f64; 16],
-    pos: u8,
-    full: bool,
+    smoothed_steps_per_sec: f64,
+    double_smoothed_steps_per_sec: f64,
     prev: (u64, Instant),
+    start: Instant,
+    exponent_weight_seconds: f64,
+    double_smoothing: bool,
 }
 
 impl Estimator {
-    fn new(now: Instant) -> Self {
+    fn new() -> Self {
+        let now = Instant::now();
         Self {
-            steps: [0.0; 16],
-            pos: 0,
-            full: false,
+            smoothed_steps_per_sec: 0.0,
+            double_smoothed_steps_per_sec: 0.0,
             prev: (0, now),
+            start: now,
+            exponent_weight_seconds: 15.0,
+            double_smoothing: true,
         }
     }
 
+    pub(crate) fn set_exponent_weight_secs(&mut self, ews: f64) {
+        self.exponent_weight_seconds = ews;
+    }
+
+    pub(crate) fn enable_double_smoothing(&mut self) {
+        self.double_smoothing = true;
+    }
+
+    pub(crate) fn disable_double_smoothing(&mut self) {
+        self.double_smoothing = false;
+    }
+
     fn record(&mut self, new: u64, now: Instant) {
-        let delta = new.saturating_sub(self.prev.0);
-        if delta == 0 || now < self.prev.1 {
+        // sanity check: don't record data if time or steps have not advanced
+        if new <= self.prev.0 || now < self.prev.1 {
             // Reset on backwards seek to prevent breakage from seeking to the end for length determination
             // See https://github.com/console-rs/indicatif/issues/480
             if new < self.prev.0 {
-                self.reset(now);
+                self.reset();
             }
             return;
         }
 
-        let elapsed = now - self.prev.1;
-        let divisor = delta as f64;
-        let mut batch = 0.0;
-        if divisor != 0.0 {
-            batch = duration_to_secs(elapsed) / divisor;
-        };
+        let delta_steps = new - self.prev.0;
```

Comment:
One trick I used in a previous job was to just straight up set the `smoothed_steps_per_sec` (and `double_smoothed_steps_per_sec`) to the new value if they were uninitialized. Maybe that would be a decent approach here? Should those values become `Option` in that case?

Reply:
I think there are two separate questions here:

1. What should the estimator return when a value is requested but it is in the uninitialized state?
2. What the should original (uninitialized) *internal* value of the estimator be?

Because the weight of data in the estimator is always positive (there's no horizon at which the weight = 0), the only sensible answer to (2) is 0. That's because you have no way to pick the right value ahead of time, so you will want to divide out the effect of the original estimator value (I call this debiasing). This is much easier to do if it is 0, because then you just divide the naive estimate by `weight(age)`.

I don't really have an opinion on (1) so I'm happy to have it do whatever you like. If you'd like to have an uninitialized estimator return `None`, that's fine with me.

BTW in case it's unclear, after the first `update` the estimator value *will* be equal to the value of the first update (for both smoothing and double smoothing), it's just that prior to the first update there's not a correct rate to return.

Reply:
> I think there are two separate questions here:
> 
> 1. What should the estimator return when a value is requested but it is in the uninitialized state?
> ..
> I don't really have an opinion on (1) so I'm happy to have it do whatever you like. If you'd like to have an uninitialized estimator return `None`, that's fine with me.

Having an unitialized estimator return `None` sounds good to me in theory at least, though it might depend a bit on how much downstream complexity that requires to make work? Off the top of my head it seems like it might not be too much and would force us to write something like `N/A` in the progress bars, which seems good?

> 2. What the should original (uninitialized) _internal_ value of the estimator be?
> 
> Because the weight of data in the estimator is always positive (there's no horizon at which the weight = 0), the only sensible answer to (2) is 0. That's because you have no way to pick the right value ahead of time, so you will want to divide out the effect of the original estimator value (I call this debiasing). This is much easier to do if it is 0, because then you just divide the naive estimate by `weight(age)`.

Fair enough.

> BTW in case it's unclear, after the first `update` the estimator value _will_ be equal to the value of the first update (for both smoothing and double smoothing), it's just that prior to the first update there's not a correct rate to return.

Okay, good!